File: 757-pyramid-transition-matrix/pyramid-transition-matrix.py
class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:

        mapping = collections.defaultdict(list)
        for pattern in allowed:
            mapping[pattern[:2]].append(pattern[2])

        def backtrack(i_cur, cur_row, upper_row):
            # print(i_cur, cur_row, upper_row)
            if len(cur_row) == 1:
                return True

            if i_cur == len(cur_row) - 1:
                return backtrack(0, upper_row, list())

            cur_key = cur_row[i_cur] + cur_row[i_cur + 1]
            if cur_key not in mapping:
                return False

            for top in mapping[cur_key]:
                upper_row.append(top)
                if backtrack(i_cur + 1, cur_row, upper_row):
                    return True
                upper_row.pop()

            if len(upper_row) > 0:
                backtrack(0, upper_row, list())

            return False

        return backtrack(0, list(bottom), list())

File: 757-pyramid-transition-matrix/pyramid-transition-matrix.java
class Solution {

    private Map<String, List<Character>> store;

    public boolean pyramidTransition(String bottom, List<String> allowed) {
        // create mapping one, two -> list(three)
        store = new HashMap<>();
        for (String s : allowed) {
            String key = s.substring(0, 2);
            Character value = s.charAt(2);
            if (!store.containsKey(key)) {
                store.put(key, new ArrayList<>());
            }
            store.get(key).add(value);
        }

        return backtrack(0, bottom, "");
    }

    private boolean backtrack(int i, String curStr, String nextStr) {
        if (i == curStr.length() - 1) {
            i = 0;
            curStr = nextStr;
            nextStr = "";
        }

        if (i == 0 && curStr.length() == 1) {
            return true;
        }

        String key = curStr.substring(i, i + 2);
        if (!store.containsKey(key)) return false;
        for (char c : store.get(key)) {
            if (backtrack(i + 1, curStr, nextStr + c)) return true;
        }
        return false;
    }
}